### PR TITLE
Issue #138 - add ability for LogCapture to fail if errors are being logged and unasserted.

### PR DESCRIPTION
**Pull request by zolyfarkas-fb**

### CODE REVIEW

File: testfixtures/logcapture.py [hunk 1]
```diff
@@ -1,12 +1,26 @@
...
+import os
 import atexit
 import logging
 import warnings
 from pprint import pformat
-
 from testfixtures.comparison import compare
 from testfixtures.utils import wrap
 
+_FAIL_ON_ERROR_LOG = os.environ.get("TEST_FIXTURES_FAIL_ON_ERROR_LOG") == "True"
```
Comment (cjw296) [inline testfixtures/logcapture.py:10]: Nothing else is controlled by environment variables and I don't want to start that pattern, so let's lose this.
If you want it to be "on everywhere", just have a subclass or partial in your project.

File: testfixtures/logcapture.py [hunk 2]
```diff
@@ -1,12 +1,26 @@
...
+def is_fail_on_error_log():  # () -> bool
+    """ check is fail on un-asserted error log mode is on """
+    global _FAIL_ON_ERROR_LOG
+    return _FAIL_ON_ERROR_LOG
+
+
+def set_fail_on_error_log(val):
+    """ Enable/Disable fail on un-asserted error log mode """
+    global _FAIL_ON_ERROR_LOG
+    _FAIL_ON_ERROR_LOG = val
```
Comment (cjw296) [inline testfixtures/logcapture.py:22]: No globals please.

File: testfixtures/logcapture.py [hunk 3]
```diff
@@ -60,7 +80,9 @@ def __init__(self, names=None, install=True, level=1, propagate=None,
         self.attributes = attributes
         self.recursive_check = recursive_check
         self.old = defaultdict(dict)
-        self.clear()
+        self.fail_on_error_log = fail_on_error_log
+        self.records = []  # List[LogRecord]
+        self.asserted = []  # List[bool]
```
Comment (cjw296) [inline testfixtures/logcapture.py:85]: Leave the call to clear as it was please.

File: testfixtures/logcapture.py [hunk 4]
```diff
@@ -94,7 +155,7 @@ def install(self):
             self.old['levels'][name] = logger.level
             self.old['handlers'][name] = logger.handlers
             self.old['disabled'][name] = logger.disabled
-            self.old['progagate'][name] = logger.propagate
+            self.old['propagate'][name] = logger.propagate
```
Comment (cjw296) [inline testfixtures/logcapture.py:158]: Nice catch, but please separate this into its own commit.

File: testfixtures/logcapture.py [hunk 5]
```diff
@@ -73,12 +95,51 @@ def atexit(cls):
...
+            if record.levelno == logging.ERROR and not self.asserted[i]:
+                raise AssertionError((
+                    'Not asserted ERROR log: %s'
+                ) % (pformat(self[i])))
+            i = i + 1
 
-    def emit(self, record):
+    def emit(self, record):  # (self, record: LogRecord) -> None
         self.records.append(record)
+        self.asserted.append(False)
```
Comment (cjw296) [inline testfixtures/logcapture.py:142]: How about just giving each record an `asserted` attribute instead of the parallel list?

File: testfixtures/logcapture.py [hunk 6]
```diff
@@ -120,12 +181,12 @@ def uninstall(self):
...
                 logger.handlers = self.old['handlers'][name]
                 logger.disabled = self.old['disabled'][name]
-                logger.propagate = self.old['progagate'][name]
+                logger.propagate = self.old['propagate'][name]
             self.instances.remove(self)
 
     @classmethod
     def uninstall_all(cls):
-        "This will uninstall all existing :class:`LogHandler` objects."
+        """This will uninstall all existing :class:`LogHandler` objects."""
```
Comment (cjw296) [inline testfixtures/logcapture.py:189]: Not sure changes like this are worth it, but if you must, please put them in their own commit.

File: testfixtures/logcapture.py [hunk 7]
```diff
@@ -225,15 +289,18 @@ def check_present(self, *expected, **kw):
             expected = list(expected)
             matched = []
             unmatched = []
+            i = 0
             for entry in actual:
```
Comment (cjw296) [inline testfixtures/logcapture.py:293]: I don't think we'll end up with the separate `asserted` list, but the way to do this in Python is:
```
for i, entry in enumerate(actual):
```

File: testfixtures/tests/test_logcapture.py [hunk 8]
```diff
@@ -25,6 +25,52 @@ def test_simple(self):
         root.info('after')
         assert str(l) == "root INFO\n  during"
 
+    def test_simple_strict(self):
+        log_capture = LogCapture(fail_on_error_log=True)
+        root.error('during')
+        log_capture.uninstall()
+        try:
```
Comment (cjw296) [inline testfixtures/tests/test_logcapture.py:32]: Use `ShouldAssert` please, and make sure you assert about the error message shown.

File: .gitignore [hunk 9]
```diff
@@ -21,4 +21,7 @@ pip-selfcheck.json
 .coverage.*
 .cache
 .pytest*
+.idea
 /build
+/requirements.txt
+/.idea/
```
Comment (cjw296) [inline .gitignore:27]: Please remove these changes:
- `.idea` belongs in your system-wide gitignore (PyCharm is your choice, not the project's ;-))
- there's not requirements.txt in this project

File: testfixtures/logcapture.py [hunk 10]
```diff
@@ -73,12 +95,51 @@ def atexit(cls):
...
+            self.asserted[i] = True
+
+    def assert_no_error(self):  # (...) -> None
+        """ raise  AssertionError if there are un-asserted ERROR logs in this capture """
+        i = 0
+        for record in self.records:
+            if record.levelno == logging.ERROR and not self.asserted[i]:
+                raise AssertionError((
+                    'Not asserted ERROR log: %s'
+                ) % (pformat(self[i])))
```
Comment (cjw296) [inline testfixtures/logcapture.py:137]: I think we want an exception that highlights *all* the messages not asserted, rather than just the first one.

**Review by cjw296 — COMMENTED** [review]
This is a good start, but I don't think we want to hard code this to be about errors. So, big things:

- the methods shouldn't have "error" in the name.
- Instead of `fail_on_error_log` to the constructor, how about `ensure_assertions_for` and have that take a log level.

Anything logged at or above that level would need to have an assertion.

Obviously, the CI needs to pass too ;-)

File: .gitignore (hunk 9 above)
Comment (zolyfarkas) [reply .gitignore:27]: That's fine, on the other hand it would make life easier to contributors who use idea...

File: testfixtures/logcapture.py (hunk 3 above)
Comment (zolyfarkas) [reply testfixtures/logcapture.py:85]: I can do that, but I don't think that it is good practice to have members declared outside __init__, It makes the code more difficult to understand. On top of that lint tools don't like it either.

File: testfixtures/logcapture.py (hunk 1 above)
Comment (zolyfarkas) [reply testfixtures/logcapture.py:10]: I would like to be able to enable strict on existing uses of LogCapture without code modifications...that is the point of this.
I can go the path of creating utility wrappers in my code, but from a user perspective it is not great.
Env var might not be the right way to tweak behaviors, is there any other acceptable way?

File: testfixtures/logcapture.py [hunk 11]
```diff
@@ -10,12 +10,14 @@
 _FAIL_ON_ERROR_LOG = os.environ.get("TEST_FIXTURES_FAIL_ON_ERROR_LOG") == "True"
 
 
-def is_fail_on_error_log():
+def is_fail_on_error_log():  # () -> bool
```
[the remark this answers is not in the document]
Comment (zolyfarkas) [reply testfixtures/logcapture.py:13]: Any reason why?

**Comment by zolyfarkas** [comment]
@cjw296 Updated the PR based on your comments, 
I still have the default for the "strict" setting,  being read from a env var, mostly to have a simple way of tuning the strict mode on without modifying code. Let me know if there is a better way to achieve this.
Also let me know about any other issues left...

**Comment by cjw296** [comment]
As I said before: nothing else is controlled by environment variables and I don't want to start that pattern
If you want it to be "on everywhere", have a subclass or partial in your project.

**Comment by zolyfarkas-fb** [comment]
@cjw296 since you feel strongly about env vars, I removed the env var. In general I think it is useful to have a method to control these kind of settings.
I can work with monkey patching ensure_assertions_for method for changing the default behavior.
thanks for the code feedback!

File: testfixtures/logcapture.py [hunk 12]
```diff
@@ -3,11 +3,18 @@
 import logging
 import warnings
 from pprint import pformat
-
```
Comment (cjw296) [inline testfixtures/logcapture.py:6]: nitpick: can you leave this whitespace where it was please.

File: testfixtures/logcapture.py [hunk 13]
```diff
@@ -3,11 +3,18 @@
...
 from testfixtures.comparison import compare
 from testfixtures.utils import wrap
 
 
+def ensure_assertions_for():
+    """
+    :return: The minimum log level assertions are required for.
+    :rtype: int
+    """
+    return logging.NOTSET
```
Comment (cjw296) [inline testfixtures/logcapture.py:15]: Let's lose this function and just have the default be static.
To support your monkeypatching plan, I'd suggest a `default_level_for_required_checks` class attribute on `LogCapture` that's set to `logging.NOTSET`.

File: testfixtures/logcapture.py [hunk 14]
```diff
@@ -50,7 +60,9 @@ class LogCapture(logging.Handler):
 
     def __init__(self, names=None, install=True, level=1, propagate=None,
                  attributes=('name', 'levelname', 'getMessage'),
-                 recursive_check=False):
+                 recursive_check=False,
+                 ensure_assertions_for_level=ensure_assertions_for()
```
Comment (cjw296) [inline testfixtures/logcapture.py:64]:
```suggestion
                 ensure_checks_above=None
```

File: testfixtures/logcapture.py [hunk 15]
```diff
@@ -60,7 +72,8 @@ def __init__(self, names=None, install=True, level=1, propagate=None,
         self.attributes = attributes
         self.recursive_check = recursive_check
         self.old = defaultdict(dict)
-        self.clear()
+        self.ensure_assertions_for_level = ensure_assertions_for_level
+        self.records = []  # List[LogRecord]
```
Comment (cjw296) [inline testfixtures/logcapture.py:76]:
```suggestion
        if ensure_checks_above is None:
            ensure_checks_above = self. default_level_for_required_checks
        self. ensure_checks_above = ensure_checks_above
        self.clear()
```

File: testfixtures/logcapture.py [hunk 16]
```diff
@@ -130,12 +185,35 @@ def uninstall_all(cls):
...
+        :rtype: Union[Tuple, Any]
+        """
+        if callable(self.attributes):
+            return self.attributes(record)
+        else:
+            result = tuple(self._actual_row(record))
+            if len(result) == 1:
+                return result[0]
+            else:
+                return result
```
Comment (cjw296) [inline testfixtures/logcapture.py:215]: I think this can all be collapsed into `_actual_row`.

File: testfixtures/logcapture.py [hunk 17]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
                 '%s' % ('\n'.join((str(i.names) for i in cls.instances)))
                 )
 
+    def __len__(self):
+        return len(self.records)
+
+    def __getitem__(self, key):
+        return self._rec_2_obj(self.records[key])
```
Comment (cjw296) [inline testfixtures/logcapture.py:93]: What is `key` here? Maybe we should use `index` for clarity? (if I understand this correctly)

File: testfixtures/logcapture.py [hunk 18]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+
+    def __contains__(self, what):
+        for i, item in enumerate(self):
+            if what == item:
+                self.records[i].is_asserted = True
+                return True
+        return False
```
Comment (cjw296) [inline testfixtures/logcapture.py:100]: nitpick: the `return False` isn't required here, default return is `None` which has the same effect.

File: testfixtures/logcapture.py [hunk 19]
```diff
@@ -145,17 +223,13 @@ def actual(self):
...
         for r in self.records:
-            if callable(self.attributes):
-                actual.append(self.attributes(r))
-            else:
-                result = tuple(self._actual_row(r))
-                if len(result) == 1:
-                    actual.append(result[0])
-                else:
-                    actual.append(result)
+            actual.append(self._rec_2_obj(r))
```
Comment (cjw296) [inline testfixtures/logcapture.py:232]:
```suggestion
            actual.append(self._actual_row(r))
```
...once `_rec_2_obj` has been collapsed into `_actual_row`

File: testfixtures/logcapture.py [hunk 20]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+                return True
+        return False
+
     def clear(self):
         """Clear any entries that have been captured."""
         self.records = []
 
+    def mark_logs_as_asserted(self):
+        """ Mark all captured logs as asserted,
+         use this when doing the assertion externally """
```
Comment (cjw296) [inline testfixtures/logcapture.py:108]: nitpick: please have the triple quotes on their own lines to match the style in the rest of this codebase.

File: testfixtures/logcapture.py [hunk 21]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+            if what == item:
+                self.records[i].is_asserted = True
+                return True
+        return False
+
     def clear(self):
         """Clear any entries that have been captured."""
         self.records = []
 
+    def mark_logs_as_asserted(self):
```
Comment (cjw296) [inline testfixtures/logcapture.py:106]:
```suggestion
    def mark_all_checked(self):
```

File: testfixtures/logcapture.py [hunk 22]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+        """ Mark all captured logs as asserted,
+         use this when doing the assertion externally """
+        for record in self.records:
+            record.is_asserted = True
```
Comment (cjw296) [inline testfixtures/logcapture.py:110]:
```suggestion
            record.checked = True
```

File: testfixtures/logcapture.py [hunk 23]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+
+    def assert_no_un_asserted_logs(self, level=None):
```
Comment (cjw296) [inline testfixtures/logcapture.py:112]:
```suggestion
    def ensure_checked(self, level=None):
```

File: testfixtures/logcapture.py [hunk 24]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+        """
+        raise  AssertionError if there are un-asserted logs with level greater or equal wit level
```
Comment (cjw296) [inline testfixtures/logcapture.py:114]:
```suggestion
        Ensure every entry logged above the specified `level` has been checked. 
        Raises an :class:`AssertionError` if this is not the case.
```

File: testfixtures/logcapture.py [hunk 25]
```diff
@@ -60,7 +72,8 @@ def __init__(self, names=None, install=True, level=1, propagate=None,
         self.attributes = attributes
         self.recursive_check = recursive_check
         self.old = defaultdict(dict)
-        self.clear()
+        self.ensure_assertions_for_level = ensure_assertions_for_level
```
Comment (cjw296) [inline testfixtures/logcapture.py:75]:
```suggestion
        #: The level above which checks must be made for logged events.
        self.ensure_checks_above = ensure_checks_above
```

File: testfixtures/logcapture.py [hunk 26]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+    def __len__(self):
+        return len(self.records)
+
+    def __getitem__(self, key):
+        return self._rec_2_obj(self.records[key])
+
+    def __contains__(self, what):
+        for i, item in enumerate(self):
+            if what == item:
+                self.records[i].is_asserted = True
```
Comment (cjw296) [inline testfixtures/logcapture.py:98]:
```suggestion
                self.records[i].checked = True
```

File: testfixtures/logcapture.py [hunk 27]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+                return True
+        return False
+
     def clear(self):
         """Clear any entries that have been captured."""
         self.records = []
 
+    def mark_logs_as_asserted(self):
+        """ Mark all captured logs as asserted,
```
Comment (cjw296) [inline testfixtures/logcapture.py:107]:
```suggestion
        """ Mark all captured events as checked,
```

File: testfixtures/logcapture.py [hunk 28]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+         use this when doing the assertion externally """
+        for record in self.records:
+            record.is_asserted = True
+
+    def assert_no_un_asserted_logs(self, level=None):
+        """
+        raise  AssertionError if there are un-asserted logs with level greater or equal wit level
+        :param level: the logging level, defaults to self.ensure_assertions_for
```
Comment (cjw296) [inline testfixtures/logcapture.py:115]:
```suggestion
        :param level: the level above which checks must be made, defaults to :attr:`ensure_checks_above`.
```

File: testfixtures/logcapture.py [hunk 29]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
+            return
+        un_asserted = []
+        for record in self.records:
+            if level <= record.levelno and not record.is_asserted:
+                un_asserted.append(self._rec_2_obj(record))
+        if un_asserted:
+            raise AssertionError((
+                    'Not asserted ERROR log(s): %s'
+                ) % (pformat(un_asserted)))
+
```
Comment (cjw296) [inline testfixtures/logcapture.py:129]: asserted -> checked in the above.

File: testfixtures/logcapture.py [hunk 30]
```diff
@@ -73,11 +86,53 @@ def atexit(cls):
...
     def emit(self, record):
+        """
+        :param record: a LogRecord instance
+        :type record: logging.LogRecord
+        """
+        record.is_asserted = False
         self.records.append(record)
```
Comment (cjw296) [inline testfixtures/logcapture.py:136]: Lose this docstring, it's not part of the public api.

File: testfixtures/logcapture.py [hunk 31]
```diff
@@ -130,12 +185,35 @@ def uninstall_all(cls):
             i.uninstall()
 
     def _actual_row(self, record):
+        """
+        Get a attribute iterator of a LogRecord.
+        :param record: a Log Record.
+        :type record: logging.LogRecord
+        :rtype: Iterator[Any]
+        """
```
Comment (cjw296) [inline testfixtures/logcapture.py:193]: Lose the docstring, same reason as above.

File: testfixtures/logcapture.py [hunk 32]
```diff
@@ -174,11 +248,13 @@ def check(self, *expected):
...
           passed to the constructor.
         """
-        return compare(
+        result = compare(
             expected,
             actual=self.actual(),
             recursive=self.recursive_check
             )
+        self.mark_logs_as_asserted()
+        return result
```
Comment (cjw296) [inline testfixtures/logcapture.py:257]: I don't think this should mark as checked. If you're relying on `str` to do checks (which is pretty much a deprecated behaviour now anyway) and you have checking for assertions turned on, then you probably want to have a manual `mark_all_checked ` call to be explicit about your intentions.

Hopefully you have a test for this and you can change it to ensure that str'ing does *not* ensure checked.

**Review by cjw296 — COMMENTED** [review]
This is looking really good!
Just tweaks to some naming and a few bits of tidy-up and it should be good to land.
I'm happy to squash this down into one commit or your can rebase into a semantic commit or two, your choice.

File: testfixtures/logcapture.py (hunk 30 above)
Comment (zolyfarkas-fb) [reply testfixtures/logcapture.py:136]: I have added these as an aid for the next dev (including myself in the future) who will look at this code. I do not see why private apis should not be documented, but if you feel strongly about it I will remove them.

Comment (cjw296) [reply testfixtures/logcapture.py:136]: I'm afraid I do: if it's a docstring, it shows up in the API docs and forms part of the public API, which I'd like to avoid for these methods.

Comments are fine to help future developers...

File: testfixtures/logcapture.py (hunk 32 above)
Comment (zolyfarkas-fb) [reply testfixtures/logcapture.py:257]: I think you need to expand this diff, the mark as checked in not part of the str method, it is part of the check method.

Comment (cjw296) [reply testfixtures/logcapture.py:247]: perfect!

File: testfixtures/tests/test_logcapture.py [hunk 33]
```diff
@@ -25,6 +26,58 @@ def test_simple(self):
...
 
+    def test_simple_strict(self):
+        log_capture = LogCapture(ensure_checks_above=ERROR)
+        root.error('during')
+        log_capture.uninstall()
+        with ShouldAssert("Not asserted ERROR log(s): [('root', 'ERROR', 'during')]"):
+            log_capture.ensure_checked()
+
+    def test_simple_strict_patched(self):
+        re_defaulted_init = partialmethod(LogCapture.__init__, ensure_checks_above=ERROR)
```
Comment (cjw296) [inline testfixtures/tests/test_logcapture.py:37]: Having the class attribute with the default level in would allow your monkeypatching to be significantly more pleasant :-)

**Comment by zolyfarkas-fb** [comment]
@cjw296 Regarding check vs assert, will change to check to be more consistent with existing nomenclature.

However I believe assert is the clearer term and might be worthwhile using in the future. To give you an example, there is a class of lint checks that validate unit test quality by looking at the implementation and looking for assertion code.(like [UTAO_JUNIT_ASSERTION_ODDITIES_NO_ASSERT](http://fb-contrib.sourceforge.net/bugdescriptions.html)) Using standard terms like assert, helps these tools recognize assertions.

Other than some clarification about doc-strings, will push out an updated PR to review.

File: testfixtures/logcapture.py (hunk 30 above)
Comment (zolyfarkas) [reply testfixtures/logcapture.py:134]: Ok, I understand now, thanks for the explanation.

**Comment by cjw296** [comment]
I'm fine with "check", and I think linting for the presence of asserts sounds like a terrible idea.

**Comment by cjw296** [comment]
Also, please can you lose the partialmethod stuff completely and just replace with a class attribute as I suggested:

```
class LogCapture(logging.Handler):
    ...
    default_level_for_required_checks = logging.NOTSET
```

Then in your code you can just do:

```
LogCapture.default_level_for_required_checks = logging.WARNING
```

...or:

```
class CheckedLogCapture(LogCapture)
    default_level_for_required_checks = logging.WARNING
```

**Comment by zolyfarkas-fb** [comment]
> I'm fine with "check", and I think linting for the presence of asserts sounds like a terrible idea.

The lint tools are looking for asserts and will emit a violation when a unit test is missing any assertions. 
Missing assertions in a unit test is a sign of a low quality test.
All the tests that have been added as part of this PR assert(or check) something. If they wouldn't I am pretty sure you would have objected :-).

**Comment by zolyfarkas-fb** [comment]
> Also, please can you lose the partialmethod stuff completely and just replace with a class attribute as I suggested:
> 
> ```
> class LogCapture(logging.Handler):
>     ...
>     default_level_for_required_checks = logging.NOTSET
> ```
> 
> Then in your code you can just do:
> 
> ```
> LogCapture.default_level_for_required_checks = logging.WARNING
> ```
> 
> ...or:
> 
> ```
> class CheckedLogCapture(LogCapture)
>     default_level_for_required_checks = logging.WARNING
> ```

Done

File: testfixtures/logcapture.py [hunk 34]
```diff
@@ -60,7 +66,12 @@ def __init__(self, names=None, install=True, level=1, propagate=None,
         self.attributes = attributes
         self.recursive_check = recursive_check
         self.old = defaultdict(dict)
-        self.clear()
```
Comment (cjw296) [inline testfixtures/logcapture.py:63]: I have requested this more than once now: please put this back.

File: testfixtures/logcapture.py [hunk 35]
```diff
@@ -60,7 +66,12 @@ def __init__(self, names=None, install=True, level=1, propagate=None,
...
+        #: The log level above which checks must be made for logged events.
+        if ensure_checks_above:
+            self.ensure_checks_above = ensure_checks_above
+        else:
+            self.ensure_checks_above = self.__class__.default_ensure_checks_above
```
Comment (cjw296) [inline testfixtures/logcapture.py:73]: Pretty sure I suggested this already, but the above is really horrible Python...
(also, the `is None` is important, otherwise `LogCapture(ensure_checks_above=0)` will not work)
```suggestion
        if ensure_checks_above is None:
            self.ensure_checks_above = self.default_ensure_checks_above
        else:
            self.ensure_checks_above = ensure_checks_above
```

File: testfixtures/logcapture.py [hunk 36]
```diff
@@ -60,7 +66,12 @@ def __init__(self, names=None, install=True, level=1, propagate=None,
...
+        self.records = []  # List[LogRecord]
```
Comment (cjw296) [inline testfixtures/logcapture.py:74]: remove this line

File: testfixtures/logcapture.py [hunk 37]
```diff
@@ -73,11 +84,54 @@ def atexit(cls):
...
+                return True
+
     def clear(self):
         """Clear any entries that have been captured."""
         self.records = []
 
+    def mark_all_checked(self):
+        """
+        Mark all captured events as checked,
+        use this when doing the assertion externally
```
Comment (cjw296) [inline testfixtures/logcapture.py:106]:
```suggestion
        Mark all captured events as checked.
        This should be called if you have made assertions about logging
        other than through :class:`LogCapture` methods.
```

File: testfixtures/logcapture.py [hunk 38]
```diff
@@ -73,11 +84,54 @@ def atexit(cls):
...
+        for record in self.records:
+            record.checked = True
+
+    def ensure_checked(self, level=None):
+        """
+        Ensure every entry logged above the specified `level` has been checked.
+        Raises an :class:`AssertionError` if this is not the case.
+
+        :param level: the logging level, defaults to :attr:`ensure_checks_above`.
+        :type level: Union[None, int]
```
Comment (cjw296) [inline testfixtures/logcapture.py:117]: If you're going to put types on...
```suggestion
        :type level: Optional[int]
```

File: testfixtures/logcapture.py [hunk 39]
```diff
@@ -73,11 +84,54 @@ def atexit(cls):
...
+        """
+        if level is None:
+            level = self.ensure_checks_above
+        if level == logging.NOTSET:
+            return
+        un_checked = []
+        for record in self.records:
+            if level <= record.levelno and not record.checked:
```
Comment (cjw296) [inline testfixtures/logcapture.py:125]: Never seen "greater than" expressed this way round before, let's change it:
```suggestion
            if record.levelno >= level and not record.checked:
```

**Review by cjw296 — CHANGES_REQUESTED** [review]

File: testfixtures/logcapture.py (hunk 34 above)
Comment (zolyfarkas) [reply testfixtures/logcapture.py:63]: Will do, however I do not agree the result is better, the maintainer has to do extra effort to understand the code.

File: testfixtures/logcapture.py (hunk 38 above)
Comment (zolyfarkas-fb) [reply testfixtures/logcapture.py:117]: I guess I authored to many avro schemas :-), will change

File: testfixtures/logcapture.py (hunk 34 above)
Comment (zolyfarkas-fb) [reply testfixtures/logcapture.py:63]: Also lint is on my side as well: "Instance attribute records defined outside __init__"

**Comment by zolyfarkas** [comment]
I have squashed all commits into one, and created a new PR. Hopefully I did not miss anything.
